aries_cloudagent/protocols/did_rotate/v1_0/manager.py

```diff
@@ -0,0 +1,244 @@
+        except ReportableDIDRotateError as err:
+            responder = self.profile.inject(BaseResponder)
+            err.message.assign_thread_from(rotate)
+            await responder.send(err.message, connection_id=conn.connection_id)
+
+        async with self.profile.session() as session:
+            await record.save(session, reason="Received rotate message")
+
+        return record
+
+    async def commit_rotate(self, conn: ConnRecord, record: RotateRecord):
+        """Commit rotate.
+
+        Args:
+            conn (ConnRecord): The connection to rotate the DID for.
+            record (RotateRecord): The rotate record.
+        """
+        record.state = RotateRecord.STATE_ACK_SENT
+        if not record.new_did:
+            raise ValueError("No new DID stored in record")
+
+        conn_mgr = BaseConnectionManager(self.profile)
+        try:
+            await conn_mgr.record_keys_for_resolvable_did(record.new_did)
+        except BaseConnectionManagerError as err:
+            # TODO Make this reportable?
+            raise DIDRotateManagerError(
+                "Unable to record keys for resolvable DID"
+            ) from err
+
+        conn.their_did = record.new_did
+
+        ack = RotateAck()
+        ack.assign_thread_id(thid=record.thread_id)
+
+        responder = self.profile.inject(BaseResponder)
+        await responder.send(ack, connection_id=conn.connection_id)
+
+        async with self.profile.session() as session:
+            # Don't emit a connection event for this change
+            # Controllers should listen for the rotate event instead
+            await conn.save(session, reason="Their DID rotated", event=False)
+            await record.save(session, reason="Sent rotate ack")
+
+        # TODO it would be better if the cache key included DIDs so we don't
+        # have to manually clear it. This is a bigger change than a first pass
+        # warrants though.
+        await conn_mgr.clear_connection_targets_cache(conn.connection_id)
+
+    async def receive_ack(self, conn: ConnRecord, ack: RotateAck):
+        """Receive rotate ack message.
+
+        Args:
+            conn (ConnRecord): The connection to rotate the DID for.
+            ack (RotateAck): The received rotate ack message.
+        """
+        async with self.profile.session() as session:
+            record = await RotateRecord.retrieve_by_thread_id(session, ack._thread_id)
+
+        record.state = RotateRecord.STATE_ACK_RECEIVED
+        if not record.new_did:
+            raise ValueError("No new DID stored in record")
+
+        conn.my_did = record.new_did
+        async with self.profile.session() as session:
+            # Don't emit a connection event for this change
+            # Controllers should listen for the rotate event instead
+            await conn.save(session, reason="My DID rotated", event=False)
+            await record.save(session, reason="Received rotate ack")
```

Comment:
Should we also delete the record?

Reply:
I lean towards considering it overkill. The record itself doesn't store much and I can't think of a reason why the controller would want to try to query them later (nor are we exposing that ability anyways).

Reply:
Replaced the `save` call with a `delete_record` method invocation.